lib/logstash/inputs/lumberjack.rb

```diff
@@ -126,4 +127,69 @@ def start_buffer_broker(output_queue)
       end
     end
   end
+
+  class EventDecoration < LogStash::Inputs::Base
+    public
+    def decorate(event)
+      super(event)
+    end
+  end # class EventDecoration
+
+  class ConnectionDecorator
+    def initialize(event_decoration, codec, connection)
+      @event_decoration = event_decoration
+      @connection = connection
+      @codec = codec
+    end
+
+    public
+    def run(&block)
+      @connection.run do |type, event|
+        case type
+        when :json; json_event(event, &block)
+        when :data; data_event(event, &block)
+        else; raise "unknown event type received"
+        end
+      end
+    end
+
+    private
+    def json_event(map, &block)
+      if map.is_a?(Array)
+        map.each { |item| json_event(item, &block) }
+      else
+        line = map.delete("line")
+        if line.nil?
+          event = LogStash::Event.new(map)
+          decorate(event)
+          yield event
+        else
+          @codec.decode(line) do |decoded|
+            decorate(decoded)
+            map.each { |k,v| decoded[k] = v }
+            yield decoded
+          end
+        end
+      end
+    end
+
+    private
+    def data_event(fields)
+      line = fields.delete("line")
+      @codec.decode(line) do |event|
+        begin
+          decorate(event)
+          fields.each { |k,v| event[k] = v; v.force_encoding(Encoding::UTF_8) }
+          yield event
+        rescue => e
+          raise e
+        end
+      end
+    end
+
+    private
+    def decorate(event)
+      @event_decoration.decorate(event)
+    end
+  end # ConnectionDecorator
```

Comment:
@jsvd I think they used to be, but I thought they were doing caching of method calls with `define_method`?